Feature: use git branches/tags as versions

becker33 opened the pull request:
Building on #24639, this allows versions to be prefixed by `git.`. If a version begins `git.`, it is treated as a git ref, and handled as git commits are starting in the referenced PR.

An exception is made for versions that are `git.develop`, `git.main`, `git.master`, `git.head`, or `git.trunk`. Those are assumed to be greater than all other versions, as those prefixed strings are in other contexts.

TODO:
- [ ] docs (could be combined with the git commit docs PR)

becker33 committed:
Refactor git versions as separate class

GitVersion inherits from VersionBase
Version remains as a wrapper that delegates to GitVersion or VersionBase
as appropriate based on the value

becker33 committed:
support non-commit git refs prefixed by 'git.'

becker33 committed:
minor refactor for develop-like git versions

becker33 committed:
residual references to Version as class

becker33 committed:
flake

becker33 commented:
@haampie you might be happy about some of the refactoring in this PR.

psakievich commented:
Ooo it does take care of my test case! Just tested it!
`nalu-wind ^trilinos@git.[hash]`

tgamblin commented on lib/spack/spack/version.py:
```diff
@@ -165,15 +169,27 @@ def __gt__(self, other):
+
+
+def Version(string):  # capitalized for backwards compatibility
+    string = str(string)  # to handle VersionBase and GitVersion types
```
This introduces a redundant copy everywhere `Version` is called (which is a lot of places) -- should check with `isinstance()` and not bother constructing a `str` if we do not need to.

CPython *can* intern strings so it might optimize this out in a lot of cases, but it doesn't always.

tgamblin commented on lib/spack/spack/version.py:
```diff
@@ -468,9 +449,128 @@ def intersection(self, other):
+
+class GitVersion(VersionBase):
+    def __init__(self, string):
+        string = str(string)  # In case we got a VersionBase or GitVersion object
```
see above -- I think this should check with `isinstance` and not copy unconditionally.

tgamblin commented on lib/spack/spack/version.py:
```diff
@@ -71,7 +71,7 @@ def coerce_versions(a, b):
     This is used to simplify comparison operations below so that we're always
     comparing things that are of the same type.
     """
```
docstring needs updating.

becker33 replied:
Tom also has an idea to represent all versions in the same triply nested tuple syntax to eliminate the need for `@coerced` that I'm quite excited about. Separate PR though.

tgamblin replied:
yeah I talked about that with Tom and it's way way faster.  Definitely separate PR though.

tgamblin commented on lib/spack/spack/solver/asp.py:
```diff
@@ -1432,7 +1432,7 @@ def key_fn(item):
 
                 known_versions = self.possible_versions[dep.name]
-                if (not dep.version.is_commit and
+                if (not isinstance(dep.version, spack.version.GitVersion) and
```
can you put a comment here saying why this doesn't matter for git versions?  I had to think about this for a minute to work out why we don't care.

tgamblin commented on lib/spack/spack/test/versions.py:
```diff
@@ -544,19 +544,19 @@ def test_get_item():
+    assert isinstance(b, VersionBase)
     assert b == Version('1_2-3')
-    assert repr(b) == "Version('1_2-3')"
+    assert repr(b) == "VersionBase('1_2-3')"
```
There should be another simple test to verify that `Version()` returns the right thing with commit, branch, and plain version inputs.

tgamblin commented on lib/spack/spack/test/versions.py:
```diff
@@ -637,6 +637,32 @@ def test_git_hash_comparisons(
 
 
+@pytest.mark.skipif(sys.platform == 'win32',
+                    reason="Not supported on Windows (yet)")
```
what's holding this back on windows?  Seems like if we have `git` in the path it should be fine.

becker33 replied:
I don't know what is holding it back, but the previous test `test_git_commit_comparisons` was marked with the same skip, and they follow almost-identical code paths so I'm pretty sure the same reasoning will apply.

tgamblin commented on lib/spack/spack/version.py:
```diff
@@ -165,15 +169,27 @@ def __gt__(self, other):
+        return True
+    elif len(string) == 40 and COMMIT_VERSION.match(string):
+        return True
+    return False
```
Doesn't it make sense to also allow `git.<commit hash>` for consistency?  I can see people trying to make commits that way based on branches and failing mysteriously.

becker33 replied:
This does allow `git.<commit hash>`. The first check is for `git.`, we only check the contents if it doesn't start with that string.

tgamblin commented on lib/spack/spack/version.py:
```diff
@@ -323,13 +315,9 @@ def satisfies(self, other):
         gcc@4.7 so that when a user asks to build with gcc@4.7, we can find
         a suitable compiler.
         """
-        self_cmp = self._cmp(other.commit_lookup)
-        other_cmp = other._cmp(self.commit_lookup)
```
It doesn't look like there is an implementation of `up_to` in `GitVersion` that preserves this behavior -- should there be?

becker33 replied:
This is the code for `satisfies`, which does have a counterpart. The code for `up_to` didn't need to change.

tgamblin commented on lib/spack/spack/version.py:
```diff
@@ -468,9 +449,128 @@ def intersection(self, other):
             return VersionList()
 
+
+class GitVersion(VersionBase):
```
So happy this is separate now.  But needs a docstring.

tgamblin commented on lib/spack/spack/version.py:
```diff
@@ -468,9 +449,128 @@ def intersection(self, other):
+        string = str(string)  # In case we got a VersionBase or GitVersion object
+        self.is_ref = False
+
+        if string.startswith('git.'):
+            string = string[4:]
+            self.is_ref = True
+
+        self.is_commit = len(string) == 40 and COMMIT_VERSION.match(string)
```
it looks like this allows commit versions to start with `git.`, but `Version()` doesn't.  Need a test to ensure that both work.

tgamblin replied:
In other version types, `string` preserves the *exact* original string, but here we lose `git.`.  I actually think it makes sense to keep that... shouldn't we?

psakievich replied:
IMO it makes sense to maintain the exact string the user specified for the spec in the version. This will make it clearer when reviewing a list of specs from `spack find`

becker33 replied:
`Version()` already does allow `git.<commit hash>`. The test I've added confirms that.

The reason I have `self.string` strip `git.` out and `str(self)` include it is that we often need the git ref alone. I can change that to be some other attribute and keep the full string in `self.string` if you prefer.

tgamblin replied:
Yeah I think it's better to be consistent with other versions and keep the original input string for output.

psakievich commented on lib/spack/spack/fetch_strategy.py:
```diff
@@ -1575,11 +1575,11 @@ def for_package_version(pkg, version):
 
     # if it's a commit, we must use a GitFetchStrategy
-    if version.is_commit and hasattr(pkg, "git"):
+    if isinstance(version, spack.version.GitVersion) and hasattr(pkg, "git"):
```
How does this behave in the case that `isinstance(version, spack.version.GitVersion) and !hasattr(pkg, "git")`? Should we add a check and raise an exception in that case?.

psakievich replied:
Also I think this will fail for `git.master` versions. The `kwargs` for the fetch strategy are hard coded to use `commit` below this.

Should add tests for the different `git.[ref types]` that are supported.

becker33 replied:
Modified it to raise if there is no 'git' attribute.

Treating branches/tags as commits wasn't going to cause a problem, it just has more conservative fetch semantics. I've changed it to separate by ref type.

psakievich commented on lib/spack/spack/solver/asp.py:
```diff
@@ -2189,7 +2189,7 @@ def build_specs(self, function_tuples):
             for spec in root.traverse():
-                if spec.version.is_commit:
+                if isinstance(spec.version, spack.version.GitVersion):
                     spec.version.generate_commit_lookup(spec.fullname)
```
The name `generate_commit_lookup` seems to be a little too specific for the git variations now supported by this PR.  It might be better to rename it or add another layer of abstraction to account for ref types such as `git.branch` and `git.tags`

psakievich reviewed (changes requested):
Looks really good. There are a few holes regarding git refs other than commits though.  I pointed out at least one in my review, but I think more tests need to be added for the various types of refs supported. I pulled this down and tested staging a `git.[branch]` locally.
```
spack stage yaml-cpp@git.master
From https://github.com/jbeder/yaml-cpp
 * branch            HEAD       -> FETCH_HEAD
==> Warning: There is no checksum on file to fetch yaml-cpp@git.master safely.
==>   Fetch anyway? [y/N] y
error: pathspec 'git.master' did not match any file(s) known to git
==> Error: All fetchers failed for spack-stage-yaml-cpp-git.master-hmbhd2gnl2342shpta7dhprcajxagslg
```

`spack solve yaml-cpp@git.master` works fine.

becker33 committed:
Add test for 'Version' wrapper

becker33 committed:
reduce redundant 'str' calls for performance

becker33 committed:
versions and docstrings

becker33 committed:
rename generate_commit_lookup -> generate_git_lookup

becker33 committed:
Modify GitVersion.string to include 'git.' prefix as appropriate

Add GitVersion.ref to get un-prefixed version
Modify method/attribute names to be ref_ not commit_

becker33 committed:
Ensure 'git.<hash>' and '<hash'> handled identically

becker33 committed:
fixup comment

becker33 committed:
improve fetch_strategy.for_package_version for git refs

becker33 committed:
eliminate checksum warning for non-commit git versions

becker33 committed:
ensure tag versions handled appropriately for older git

becker33 commented:
@tgamblin @psakievich I believe I've addressed all review comments.

becker33 committed:
flake

becker33 committed:
isort

becker33 committed:
isort

becker33 committed:
do not allow packages with git commits as versions

psakievich commented on lib/spack/spack/version.py:
```diff
@@ -1113,11 +1218,11 @@ def lookup_commit(self, commit):
+            # This will raise a ProcessError if the ref does not exist
             # We may later design a custom error to re-raise
-            self.fetcher.git('cat-file', '-e', '%s^{commit}' % commit)
+            self.fetcher.git('cat-file', '-e', '%s^{commit}' % ref)
```
I was playing around with this a little more, and I think it might be a good time to add a custom error. If I knowingly give it a wrong git ref (but a not unreasonable one) we get a rather unclear error message:

```console
$ spack stage yaml-cpp@git.develop
From https://github.com/jbeder/yaml-cpp
 * branch            HEAD       -> FETCH_HEAD
fatal: Not a valid object name develop^{commit}
==> Error: Command exited with status 128:
'/usr/bin/git' '-c' 'advice.detachedHead=false' 'cat-file' '-e' 'develop^{commit}'
```

I don't feel like this is a show stopper though so I won't request holding up the PR over it.  However, I think if not in this PR then soon. 🙂

becker33 replied:
```
$ spack stage yaml-cpp@git.develop
==> Error: develop is not a valid git ref for yaml-cpp
```
better?

psakievich replied:
💯

psakievich reviewed (dismissed):
LGTM

becker33 committed:
better error message for invalid refs

trws commented on lib/spack/spack/fetch_strategy.py:
```diff
@@ -1575,16 +1575,30 @@ def for_package_version(pkg, version):
     check_pkg_attributes(pkg)
 
-    if not isinstance(version, spack.version.Version):
+    if not isinstance(version, spack.version.VersionBase):
```
Missed it was a subclass.

becker33 replied:
GitVersion inherits from VersionBase

trws replied:
Yup, that's why I resolved it and added the comment to that effect.

becker33 committed:
error message for git versions without git attribute

tgamblin commented on lib/spack/spack/directives.py:
```diff
@@ -330,6 +330,16 @@ def _execute_version(pkg):
+                                 for argname, value in kwargs.items())
+                msg += ")"
+                raise VersionLookupError(msg)
         pkg.versions[Version(ver)] = kwargs
```
```python
pkg.versions[version] = kwargs
```
